File: queue/QueueImpByArray.java
package org.q10viking.queue;

//Java program for array implementation of queue

public class QueueImpByArray {
	int array[];
	int capacity;
	int front,rear; 
	int size;
	
	public QueueImpByArray(int capacity) {
		this.capacity = capacity;
		front = this.size = 0;
		rear = capacity-1;
		array = new int[capacity];
	}
	
	
	//Queue is full when size becomes equal to the capacity
	boolean isFull() {
		return size==capacity;
	}
	
	
	//Queue is empty when size is 0
	boolean isEmpty() {
		return size==0;
	}
	
	//Method to add an element to the queue
	//It change rear and size
	void enqueue(int item) {
		if(isFull()) {
			return;
		}
		//循环利用
		this.rear = (this.rear+1)%this.capacity;
		array[rear]=item;
		this.size += 1;
	}
	
	
	//Method to remove an item from queue
	//It change front and size
	int dequeue() {
		if(isEmpty()) {
			return Integer.MIN_VALUE;
		}
		int item = array[front];
		this.front = (this.front+1)%capacity;
		this.size -= 1;
		return item;
	}
	
	
	int front() {
		if(isEmpty()) {
			return Integer.MIN_VALUE;
		}
		return array[front];
	}
	
	int rear() {
		if(isEmpty()) {
			return Integer.MIN_VALUE;
		}
		return array[rear];
	}
	
	public static void main(String[] args) {
		QueueImpByArray queue = new QueueImpByArray(4);
		queue.enqueue(10);
		queue.enqueue(20);
		queue.enqueue(30);
		queue.enqueue(40);
		
		queue.enqueue(50);
		System.out.println("front: "+queue.front()+" rear: "+queue.rear());
		queue.dequeue();
		queue.enqueue(50);
		System.out.println("front: "+queue.front()+" rear: "+queue.rear());
	}

}

/*
front: 10 rear: 40
front: 20 rear: 50

*/

File: queue/QueueImpByArray.py
'''
Created on 2018年8月4日
https://www.geeksforgeeks.org/queue-set-1introduction-and-array-implementation/
@author: Viking
'''


# python3 program for array implementation of queue

# class Queue to represent a queue

class Queue:
     # __init__ function

    def __init__(self, capacity):
        self.front = self.size = 0
        self.rear = capacity - 1
        self.Q = [None] * capacity
        self.capacity = capacity

    def isFull(self):
        return self.size == self.capacity

    def isEmpty(self):
        return self.size == 0

    def Enqueue(self, item):
        if self.isFull():
            return
        self.rear = (self.rear + 1) % self.capacity
        self.Q[self.rear] = item
        self.size = self.size + 1

    def Dequeue(self):
        if self.isEmpty():
            return
        item = self.Q[self.front]
        self.front = (self.front + 1) % self.capacity
        self.size = self.size - 1
        return item

    def que_front(self):
        if self.isEmpty():
            print("Queue is Empty")
        else:
            print("Front item is", self.Q[self.front])

    def que_rear(self):
        if self.isEmpty():
            print("Queue is Empty")
        else:
            print("Rear item is", self.Q[self.rear])


# Driver Code
if __name__ == '__main__':
    queue = Queue(3)
    queue.Enqueue(10)

    queue.Enqueue(20)
    queue.Enqueue(30)
    queue.Enqueue(40)

    queue.que_front()
    queue.que_rear()

'''
Front item is 10
Rear item is 30
'''